### ISSUE TITLE
peak-finders

### ISSUE THREAD

**Issue by msdubrovin** [scikit-beam/scikit-beam#480]
Dear ALL,
For LCLS imaging detector data processing we have developed a few useful algorithms which may represent common interest. Currently code is available in different from scikit-beam packaging/release environment. We are going to begin with implementation in scikit-beam a couple of peak-finders.
Interface for this project is described in
https://confluence.slac.stanford.edu/display/PSDMInternal/Image+processing+algorithms+for+scikit
We will appreciate your comments regarding this project.

**Comment by danielballan** [comment]
Welcome @msdubrovin! Thanks for seeking input from the scikit-beam community at this stage. 

Here's my "two cents."  I'm imagining a user who wants to reuse your code but for some slightly different application. We want to make it as easy as possible for that user to repurpose the parts they need; we don't them to be tempted to reimplement or copy/paste your code to pull out the sections they are interested in. With that in mind, here are my suggestions:

The inputs look good, but I would simplify `data` by accepting just a single 2D array for processing. A peak-finding algorithm only *needs* to know about one image at time, so we might as well keep the interface and the implementation simple. If the users have several 2D array (say, panels of a detector) they will have to manage that geometry anyway, and we might as well keep that separate. Tooling to help with this could always be added in a higher layer that calls this function.

Your dictionary of outputs is a very nice, comprehensive list of useful peak characterizations. Some of these may be expensive to compute, and not all users will be interested in all of the characterizations. Therefore, I suggest expressing this in a second, separate function. Users who would prefer a single function can easily combine them in a convenience function.

So, what to output instead? There are two ways to express a set of regions:
* a list of indexes like `[(peak1_ys, peak1_xs), (peak2_ys, peak2_xs), ...]` (similar to the last part of your suggested output)
* a "label array" -- an array the shape of the input image with 0s filling the background, 1s filling the region of the first peak, then 2s, 3s, etc.

Scikit-beam already includes functions for converting between these two representations. A list is usually more efficient for sparse peaks, a label array better for dense peaks (though it depends on what you plan to do with the data next). I would suggest that peak-finders should return one of these -- whichever can be obtained more efficiently. Additional characterizations can then be performed on demand, as a separate step.

Again, additional functions (perhaps also included in scikit-beam) can provide whatever outputs are convenient, but factoring the core algorithm like this maximizes reusability and makes the code performant in a range of different applications by not computing something that might not be used. What do you think?

Sorry for the long essay....

**Comment by msdubrovin** [comment]
Daniel,
Thank you for time and thoughts about this project. 
Below I'll try to respond to your comments.

1. Using input data as a single 2D (numpy) array is included as a #1 option for this parameter.
Just in order to make these methods re-usable and convenient for LCLS users we have to
allow n-d arrays. In our experiments we have a few composite detectors which produce 3- or 4- dimensional arrays, for example CSPAD has a reasonable shape as (4,8,185,388) or (32,185,388).
Processing of n-d array is more precise and efficient way than processing of assembled image made of these array which in our cases may have up to 20-25% of unused pixels.
So parameter data option #3 is kind of immediate compatibility for our users who work with more than 2-d detector data. 
Parameter data option #2 is also requested by our users. Sometimes they do not care about entire
image, but would like to deal with smaller region of interest (ROI). So, list of these ROIs looks as a reasonable format. 
The amount of code to deal with these various data formats is probably smaller than length of this explanation. And finally, it looks convenient to have it implemented in the same method rather than
overload this implementation on users or other wrapping method.

2. dictionary of outputs;
There is an advantage of using dictionary of arrays of simple types like int or float, assuming that these arrays are contiguous in memory that allow fast loop. The first portion of the output dictionary is about these parameters. However, it is not a case for composite objects. So, in order to simplify as much as possible structure of output info, we would prefer to get through the dictionary just a list of peak objects. Then, each object may have composite data associated with peak, for example 1-d or 2-d list of indexes associated with peak,  a "label array", etc. plus additional methods which work on index arrays and using data. Access to data is not so much different in our and your cases. 
By the way, in our algorithms  a "label array" is naturally used to mark busy pixels, so it is not a problem to return it along with a few other useful (shaped as data) arrays through the peak object, but each of these arrays increases a size of result. The most compact way is to return list of pixel indexed for each peak which, as you've said, can be converted to the map. 

In summary, I think that your suggestions do not contradict to original project and well inscribed in it.
I have reflected your suggestions in the project description. 
Thanks again for comments!

**Comment by danielballan** [comment]
1. Would it be possible to factor options #2 and #3 as higher-level functions that wrap a more single-purpose function that takes in only option #1? If so, I would say that results in simpler, more approachable code.

2. I'm still not sure I understand the case for a Peak object. Code that consumes this output will need to know the names of the peak object and unpack them. Plain dicts and numpy arrays are a universal interface, and can be directly consumed by other libraries. Can you explain more the advantages of a Peak object in this case?

**Comment by msdubrovin** [comment]
Daniel,

Thank you for comments. Let me try to answer.

>> 1. Would it be possible to factor ...
By "to factor" you probably mean duplication of methods with different names. Why do you need that?  
All of suggested data types can be easily processed by the same method. There is no complication in code. Time delay for data type check is negligible comparing to time consumed by algorithms. 
2-d arrays (option #1) will be processed without any extra loops over segments etc.
On the other hand, we have to care about LCLS users. I guess, more than 80% of LCLS experiments use detectors like 
   PNCCD, data.shape=(4, 512, 512)
   CSPAD , data.shape=(32, 185, 388) or (4, 8, 185, 388)
   CSPAD2x2, data.shape=(2, 185, 388) 
and one would expect more composite detectors like that. So, for us option #2,3 is not an unusual case. This is our regular data processing technique. Moreover, in some of LCLS applications (not mine) 3-d arrays are used even for 2-d detectors in order to generalize data processing without re-shaping.

>> 2. I'm still not sure I understand the case for a Peak object...
For my perception the type "list of peak objects" sounds much better than "list of 2-d numpy arrays with peak indexes". As I suggested earlier, entire result of these methods can be returned as a "list of peak objects". Peak object structure can be described the same way as dictionary. Peak object can be derived for free to re-do peak data processing with new ideas. Peak object may contain results packed "peak-wise". In contrast, dictionary returns results packed "parameter-wise". Both cases have a right for existence. In different data processing schemes one or another can be more optimal. Dictionary for data of simple types has some advantage. However, numpy.array is not a simple type. This is another object. In this sense, there is no difference in returning list of peak objects or list of numpy arrays.
To me it is not a principal issue, so I have changed API description as you suggested.

**Comment by ordirules** [comment]
Just curious what the status here is, this sounds like a useful algorithm :-)

**Comment by msdubrovin** [comment]
Dear Julien,
There is some progress in this project. I hope that found quite natural way of how to use C++ code wrapped in cython and python in scikit-beam. Code of the peak finding algorithms already available in LCLS software. It needs to be transferred with appropriate modifications to scikit-beam. Unfortunately, currently I am under pressure of a few other urgent LCLS projects. Still hope to find some time to work on this project.

**Comment by ordirules** [comment]
ok, no worries, thanks for the update!
